index.js

```diff
@@ -36,25 +38,15 @@ internals.endsWith = function (str, suffix) {
 internals.findRule = function (domain) {
 
   var punyDomain = Punycode.toASCII(domain);
-  return internals.rules.reduce(function (memo, rule) {
-
-    if (rule.punySuffix === -1) {
-      rule.punySuffix = Punycode.toASCII(rule.suffix);
-    }
-    if (!internals.endsWith(punyDomain, '.' + rule.punySuffix) && punyDomain !== rule.punySuffix) {
-      return memo;
+  var punyDomainChunks = punyDomain.split('.');
+  for (var i = 0; i < punyDomainChunks.length; i++) {
+    var suffix = punyDomainChunks.slice(i).join('.');
+    var matchingRules = rulesByPunySuffix.get(suffix);
+    if (matchingRules) {
+      return matchingRules;
     }
-    // This has been commented out as it never seems to run. This is because
-    // sub tlds always appear after their parents and we never find a shorter
-    // match.
-    //if (memo) {
-    //  var memoSuffix = Punycode.toASCII(memo.suffix);
-    //  if (memoSuffix.length >= punySuffix.length) {
-    //    return memo;
-    //  }
-    //}
-    return rule;
-  }, null);
+  }
+  return null;
```

Comment:
I also tried to think about this one, I've read about how sometimes inverted lookup is supposed to be faster but I don't really know, LGTM!